subsys/net/lib/mqtt_socket/mqtt.c

```diff
@@ -0,0 +1,699 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/** @file mqtt.c
+ *
+ * @brief MQTT Client API Implementation.
+ */
+
+#define LOG_MODULE_NAME net_mqtt
+#define NET_LOG_LEVEL CONFIG_MQTT_LOG_LEVEL
+
+#include <net/mqtt.h>
+
+#include "mqtt_transport.h"
+#include "mqtt_internal.h"
+#include "mqtt_os.h"
+
+/** Number of buffers needed per client - one for RX and one for TX */
+#define BUFFER_COUNT_PER_CLIENT 2
+
+/** Total number of buffers needed by the module. */
+#define TOTAL_BUFFER_COUNT (BUFFER_COUNT_PER_CLIENT * MQTT_MAX_CLIENTS)
+
+/** MQTT Client table. */
+static struct mqtt_client *mqtt_client[MQTT_MAX_CLIENTS];
```

Comment:
Oh, so all this is for mqtt_live() to not take a param? Well, how about it taking a param and removing O(n) arrays?